README.md
# Do not use this repository - it's obsolete

Please follow: [https://github.com/ruzickap/action-broken-link-checker](https://github.com/ruzickap/action-broken-link-checker)

Everything here is just for **testing**...

## GitHub Actions: Broken link checker ✔

[![GitHub Marketplace](https://img.shields.io/badge/Marketplace-Broken%20Link%20Checker-blue.svg?colorA=24292e&colorB=0366d6&style=flat&longCache=true&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA4AAAAOCAYAAAAfSC3RAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAM6wAADOsB5dZE0gAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAERSURBVCiRhZG/SsMxFEZPfsVJ61jbxaF0cRQRcRJ9hlYn30IHN/+9iquDCOIsblIrOjqKgy5aKoJQj4O3EEtbPwhJbr6Te28CmdSKeqzeqr0YbfVIrTBKakvtOl5dtTkK+v4HfA9PEyBFCY9AGVgCBLaBp1jPAyfAJ/AAdIEG0dNAiyP7+K1qIfMdonZic6+WJoBJvQlvuwDqcXadUuqPA1NKAlexbRTAIMvMOCjTbMwl1LtI/6KWJ5Q6rT6Ht1MA58AX8Apcqqt5r2qhrgAXQC3CZ6i1+KMd9TRu3MvA3aH/fFPnBodb6oe6HM8+lYHrGdRXW8M9bMZtPXUji69lmf5Cmamq7quNLFZXD9Rq7v0Bpc1o/tp0fisAAAAASUVORK5CYII=)](https://github.com/ruzickap/action-broken-link-checker)
[![license](https://img.shields.io/github/license/ruzickap/action-broken-link-checker.svg)](https://github.com/ruzickap/action-broken-link-checker/blob/master/LICENSE)
[![release](https://img.shields.io/github/release/ruzickap/action-broken-link-checker.svg)](https://github.com/ruzickap/action-broken-link-checker/releases/latest)
[![GitHub release date](https://img.shields.io/github/release-date/ruzickap/action-broken-link-checker.svg)](https://github.com/ruzickap/action-broken-link-checker/releases)
![GitHub Actions status](https://github.com/ruzickap/action-broken-link-checker/workflows/docker-image/badge.svg)
[![Docker Hub Build Status](https://img.shields.io/docker/cloud/build/peru/broken-link-checker.svg)](https://hub.docker.com/r/peru/broken-link-checker)

This is a GitHub Action to check broken link in your static files or web pages.
The [muffet](https://github.com/raviqqe/muffet) is used for URL checking task.

See the basic GitHub Action example to run periodic checks (weekly)
against [google.com](https://google.com):

```yaml
on:
  schedule:
    - cron: '0 0 * * 0'

name: Check markdown links
jobs:
  broken-link-checker:
    name: Check broken links
    runs-on: ubuntu-18.04
    steps:
    - name: Check
      uses: ruzickap/action-broken-link-checker@v1
      with:
        url: https://www.google.com
        cmd_params: "--one-page-only"  # Check just one page
```

This deploy action can be combined with [Static Site Generators](https://www.staticgen.com/)
(Hugo, MkDocs, Gatsby, GitBook, mdBook, etc.). The following examples expects
to have the web page stored in `./build` directory. There is a [caddy](https://caddyserver.com/)
web server started during the tests which is using the hostname from the `URL`
parameter and serving the web pages (see the details in [entrypoint.sh](./entrypoint.sh)).

```yaml
- name: Check
  uses: ruzickap/action-broken-link-checker@v1
  with:
    url: https://www.example.com/test123
    pages_path: ./build/
    cmd_params: --buffer-size=8192 --concurrency=10 --skip-tls-verification --limit-redirections=5 --timeout=20  # muffet parameters
    run_timeout: 600  # maximum amount of time to run muffet (default is set to 300 seconds)
```

Do you want to skip the docker build step? OK, the script mode is also available:

```yaml
- name: Check
  env:
    INPUT_URL: https://www.example.com/test123
    INPUT_PAGES_PATH: ./build/
    INPUT_CMD_PARAMS: --buffer-size=8192 --concurrency=10 --skip-tls-verification  # --skip-tls-verification is mandatory parameter when using https and "PAGES_PATH"
  run: wget -qO- https://raw.githubusercontent.com/ruzickap/action-broken-link-checker/v1/entrypoint.sh | bash
```

## Parameters

Environment variables used by `./entrypoint.sh` script.

| Variable            | Default                               | Description                                                                                                                                                              |
|---------------------|---------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `INPUT_CMD_PARAMS`  | `--buffer-size=8192 --concurrency=10` | Command line parameters for URL checker [muffet](https://github.com/raviqqe/muffet) - details [here](https://github.com/raviqqe/muffet/blob/master/arguments.go#L16-L34) |
| `INPUT_DEBUG`       | false                                 | Enable debug mode for the `./entrypoint.sh` script (`set -x`)                                                                                                            |
| `INPUT_PAGES_PATH`  |                                       | Relative path to the directory with local web pages                                                                                                                      |
| `INPUT_RUN_TIMEOUT` | 300                                   | Maximum number of seconds that URL checker can be running                                                                                                                |
| `INPUT_URL`         | (**Mandatory / Required**)            | URL which will be checked                                                                                                                                                |

## Full example

GitHub Action example:

```yaml
name: Checks

on:
  push:
    branches:
    - master

jobs:
  build-deploy:
    runs-on: ubuntu-18.04
    steps:
      - name: Create web page
        run: |
          mkdir -v public
          cat > public/index.html << EOF
          <!DOCTYPE html>
          <html>
            <head>
              My page which will be stored on my-testing-domain.com domain
            </head>
            <body>
              Links:
              <ul>
                <li><a href="https://my-testing-domain.com">https://my-testing-domain.com</a></li>
                <li><a href="https://my-testing-domain.com:443">https://my-testing-domain.com:443</a></li>
              </ul>
            </body>
          </html>
          EOF

      - name: Check links using script
        env:
          INPUT_URL: https://my-testing-domain.com
          INPUT_PAGES_PATH: ./public/
          INPUT_CMD_PARAMS: "--skip-tls-verification --verbose"
          INPUT_RUN_TIMEOUT: 100
          INPUT_DEBUG: true
        run: wget -qO- https://raw.githubusercontent.com/ruzickap/action-broken-link-checker/v1/entrypoint.sh | bash

      - name: Check links using container
        uses: ruzickap/action-broken-link-checker@v1
        with:
          url: https://my-testing-domain.com
          pages_path: ./public/
          cmd_params: "--skip-tls-verification --verbose"
          run_timeout: 10
          debug: true
```

## Best practices

Let's try to automate the creating the web pages as much as possible.

The ideal situation require the repository naming convention, where the name of
the GitHub repository should match the URL where it will be hosted.

### GitHub Pages with custom domain

The mandatory part is the repository name `awsug.cz` which is the same as the
domain:

* Repository name: [awsugcz/awsug.cz](https://github.com/awsugcz/awsug.cz)
  \-> Web pages: [https://awsug.cz](https://awsug.cz)

The web pages will be stored as Github Pages on it's [own domain](https://help.github.com/en/github/working-with-github-pages/configuring-a-custom-domain-for-your-github-pages-site).

The GH Action file may looks like:

```yaml
name: hugo-build

on:
  pull_request:
    types: [opened, synchronize]
  push:

jobs:
  hugo-build:
    runs-on: ubuntu-18.04
    steps:
      - uses: actions/checkout@v2

      - name: Checkout submodules
        shell: bash
        run: |
          auth_header="$(git config --local --get http.https://github.com/.extraheader)"
          git submodule sync --recursive
          git -c "http.extraheader=$auth_header" -c protocol.version=2 submodule update --init --force --recursive --depth=1

      - name: Setup Hugo
        uses: peaceiris/actions-hugo@v2.3.2
        with:
          hugo-version: '0.62.0'

      - name: Build
        run: |
          hugo --gc
          cp LICENSE README.md public/
          echo "${{ github.event.repository.name }}" > public/CNAME

      - name: Check broken links
        env:
          INPUT_URL: https://${{ github.event.repository.name }}
          INPUT_PAGES_PATH: public
          INPUT_CMD_PARAMS: --verbose --buffer-size=8192 --concurrency=10 --skip-tls-verification --exclude=linkedin.com
        run: |
          wget -qO- https://raw.githubusercontent.com/ruzickap/action-broken-link-checker/v1/entrypoint.sh | bash

      - name: Deploy
        uses: peaceiris/actions-gh-pages@v2.8.0
        if: github.event_name == 'push' && github.ref == 'refs/heads/master'
        env:
          ACTIONS_DEPLOY_KEY: ${{ secrets.ACTIONS_DEPLOY_KEY }}
          PUBLISH_BRANCH: gh-pages
          PUBLISH_DIR: public
        with:
          forceOrphan: true
```

The example is using [Hugo](https://gohugo.io/).

### GitHub Pages with [github.io](https://github.io) domain

The mandatory part is the repository name `k8s-harbor` which is the directory
part at the and of `ruzickap.github.io`:

* Repository name: [ruzickap/k8s-harbor](https://github.com/ruzickap/k8s-harbor)
  \-> Web pages: [https://ruzickap.github.io/k8s-harbor](https://ruzickap.github.io/k8s-harbor)

In the example the web pages will be using GitHub's domain [github.io](https://github.io).

```yaml
name: vuepress-build-check-deploy

on:
  pull_request:
    types: [opened, synchronize]
    paths:
      - .github/workflows/vuepress-build-check-deploy.yml
      - docs/**
      - package.json
      - package-lock.json
  push:
    paths:
      - .github/workflows/vuepress-build-check-deploy.yml
      - docs/**
      - package.json
      - package-lock.json

jobs:
  vuepress-build-check-deploy:
    runs-on: ubuntu-18.04
    steps:
      - uses: actions/checkout@v2

      - name: Install Node.js 12
        uses: actions/setup-node@v1
        with:
          node-version: 12.x

      - name: Install VuePress and build the document
        run: |
          npm install
          npm run build
          cp LICENSE docs/.vuepress/dist
          sed -e "s@(part-@(https://github.com/${GITHUB_REPOSITORY}/tree/master/docs/part-@" -e 's@.\/.vuepress\/public\/@./@' docs/README.md > docs/.vuepress/dist/README.md

      - name: Check broken links
        env:
          INPUT_URL: https://${{ github.event.repository.owner.name }}.github.io/${{ github.event.repository.name }}
          INPUT_PAGES_PATH: .
          INPUT_CMD_PARAMS: --buffer-size=8192 --concurrency=10 --skip-tls-verification
        run: |
          ln -s docs/.vuepress/dist ${{ github.event.repository.name }}
          wget -qO- https://raw.githubusercontent.com/ruzickap/action-broken-link-checker/v1/entrypoint.sh | bash

      - name: Deploy
        uses: peaceiris/actions-gh-pages@v2.8.0
        if: github.event_name == 'push' && github.ref == 'refs/heads/master'
        env:
          ACTIONS_DEPLOY_KEY: ${{ secrets.ACTIONS_DEPLOY_KEY }}
          PUBLISH_BRANCH: gh-pages
          PUBLISH_DIR: ./docs/.vuepress/dist
        with:
          forceOrphan: true
```

In this case I'm using [VuePress](https://vuepress.vuejs.org/) to create my
page.

---

Both examples can be used as a **generic template**, and you do not need to
change them for your projects.

## Running locally

It's possible to use the checking script locally. It will install [caddy](https://caddyserver.com/)
and [muffet](https://github.com/raviqqe/muffet) binaries if they
are not already installed on your system.

```bash
export INPUT_URL="https://google.com"
export INPUT_CMD_PARAMS="--ignore-fragments --one-page-only --concurrency=10 --verbose"
./entrypoint.sh
```

Output:

```text
*** INFO: [2019-12-30 14:53:54] Start checking: "https://google.com"
https://www.google.com/
        200     http://www.google.cz/history/optout?hl=cs
        200     http://www.google.cz/intl/cs/services/
        200     https://accounts.google.com/ServiceLogin?hl=cs&passive=true&continue=https://www.google.com/
        200     https://drive.google.com/?tab=wo
        200     https://mail.google.com/mail/?tab=wm
        200     https://maps.google.cz/maps?hl=cs&tab=wl
        200     https://news.google.cz/nwshp?hl=cs&tab=wn
        200     https://play.google.com/?hl=cs&tab=w8
        200     https://www.google.com/advanced_search?hl=cs&authuser=0
        200     https://www.google.com/images/branding/googlelogo/1x/googlelogo_white_background_color_272x92dp.png
        200     https://www.google.com/intl/cs/about.html
        200     https://www.google.com/intl/cs/ads/
        200     https://www.google.com/intl/cs/policies/privacy/
        200     https://www.google.com/intl/cs/policies/terms/
        200     https://www.google.com/language_tools?hl=cs&authuser=0
        200     https://www.google.com/preferences?hl=cs
        200     https://www.google.com/setprefdomain?prefdom=CZ&prev=https://www.google.cz/&sig=K_WmKyDZc24PJiXFyTjsUeLLrG-P4%3D
        200     https://www.google.com/textinputassistant/tia.png
        200     https://www.google.cz/imghp?hl=cs&tab=wi
        200     https://www.google.cz/intl/cs/about/products?tab=wh
        200     https://www.youtube.com/?gl=CZ&tab=w1
*** INFO: [2019-12-30 14:53:55] Checks completed...
```

You can also use the advantage of the container to run the checks locally
without touching your system:

```bash
export INPUT_URL="https://google.com"
export INPUT_CMD_PARAMS="--ignore-fragments --one-page-only --concurrency=10 --verbose"
docker run --rm -t -e INPUT_URL -e INPUT_CMD_PARAMS peru/broken-link-checker
```

Another example when checking the the web page locally stored on your disk.
In this case I'm using the web page created in the `./tests/` directory from
this git repository:

```bash
export INPUT_URL="https://my-testing-domain.com"
export INPUT_PAGES_PATH="${PWD}/tests/"
export INPUT_CMD_PARAMS="--skip-tls-verification --verbose"
./entrypoint.sh
```

Output:

```text
*** INFO: Using path "/home/pruzicka/git/action-broken-link-checker/tests/" as domain "my-testing-domain.com" with URI "https://my-testing-domain.com"
*** INFO: [2019-12-30 14:54:22] Start checking: "https://my-testing-domain.com"
https://my-testing-domain.com/
        200     https://my-testing-domain.com
        200     https://my-testing-domain.com/run_tests.sh
        200     https://my-testing-domain.com:443
        200     https://my-testing-domain.com:443/run_tests.sh
https://my-testing-domain.com:443/
        200     https://my-testing-domain.com
        200     https://my-testing-domain.com/run_tests.sh
        200     https://my-testing-domain.com:443
        200     https://my-testing-domain.com:443/run_tests.sh
*** INFO: [2019-12-30 14:54:22] Checks completed...
```

The same example as above, but in this case I'm using the container:

```bash
export INPUT_URL="https://my-testing-domain.com"
export INPUT_PAGES_PATH="${PWD}/tests/"
export INPUT_CMD_PARAMS="--skip-tls-verification --verbose"
docker run --rm -t -e INPUT_URL -e INPUT_CMD_PARAMS -e INPUT_PAGES_PATH -v "$INPUT_PAGES_PATH:$INPUT_PAGES_PATH" peru/broken-link-checker
```

## Examples

Some other examples of building and checking web pages using [Static Site Generators](https://www.staticgen.com/)
and GitHub Actions can be found here: [https://github.com/peaceiris/actions-gh-pages/](https://github.com/peaceiris/actions-gh-pages/)

The following links contains real examples of Broken link checker usage:

* Go to "Actions -> check-broken-links" in this repository

* Static page generated by [Hugo](https://gohugo.io/)
  with checked links: [example](https://github.com/ruzickap/xvx.cz/blob/620bd0448113258079144b72f9a20f7a783e7791/.github/workflows/hugo-build.yml)

* Static page generated by [VuePress](https://vuepress.vuejs.org/)
  with checked links: [example](https://github.com/ruzickap/k8s-harbor/blob/0c51b62b1ad7d462268443f46c3a07cdab26fbd2/.github/workflows/vuepress-build-check-deploy.yml)

* Test 3

* Test 4


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/ruzickap/action-test/compare/v2.11.3-0...HEAD)

## [v2.11.3-0](https://github.com/ruzickap/action-test/compare/v2.11.2...v2.11.3-0) (2020-06-21)

- CHANGELOG file update [`#25`](https://github.com/ruzickap/action-test/pull/25)
- var [`c359bb7`](https://github.com/ruzickap/action-test/commit/c359bb7a5b2dc2a97728d05e66302c7c4d6e32bc)
- i [`c42ec82`](https://github.com/ruzickap/action-test/commit/c42ec82661274f034f6524665aef66bc979ff421)
- source [`ea9cf3d`](https://github.com/ruzickap/action-test/commit/ea9cf3d8c271b1e9248e9ad61a949f5b033c958e)
- grep [`47fed89`](https://github.com/ruzickap/action-test/commit/47fed89be57b95c5a30613ff6272e61bedb7cb6d)
- force_apt_get [`a78cfd4`](https://github.com/ruzickap/action-test/commit/a78cfd437de03151147423b1b63f625e53de3f10)
- Add python3 2 [`5453ec2`](https://github.com/ruzickap/action-test/commit/5453ec26cfa654b706b7906148158f40af9984e0)
- Add python3-apt [`45f8f34`](https://github.com/ruzickap/action-test/commit/45f8f34dfb87cb80ed8bcbf2f8f724dd96f0dc65)
- Ansible debug [`7b5cd76`](https://github.com/ruzickap/action-test/commit/7b5cd76373c2e25da4b9c467d4db81a6fc3589f7)
- Add become [`1de74b7`](https://github.com/ruzickap/action-test/commit/1de74b7334e3901d2c5e32a44afe95d1803752c7)
- Ansible fix [`ffc24e7`](https://github.com/ruzickap/action-test/commit/ffc24e76fe8629b6cdcff049cb33a248d622c8dd)
- Add ansible [`a679572`](https://github.com/ruzickap/action-test/commit/a6795722777bcb7e23b8b9c5926a70556aa6be96)
- shell [`3adc29f`](https://github.com/ruzickap/action-test/commit/3adc29f299e0b4a658df34b14993234a2c26a8aa)
- single [`e58a0ac`](https://github.com/ruzickap/action-test/commit/e58a0ac37170e5f2848c19f4caeddae25e7a5996)
- zzzz [`5cce950`](https://github.com/ruzickap/action-test/commit/5cce9508048d212168596ed1a84f6a66e82e5dda)
- mezery [`a83cf4c`](https://github.com/ruzickap/action-test/commit/a83cf4c41f3118b8ce4380aa1b16d3cab31aa657)
- oo [`6148e8d`](https://github.com/ruzickap/action-test/commit/6148e8dea0120eaea53bbc23b5bd0681ef20e5c9)
- aaaaxxxx [`08a923c`](https://github.com/ruzickap/action-test/commit/08a923c0690f117ae8f0242d83a6f2755bd13349)
- isss [`47f0036`](https://github.com/ruzickap/action-test/commit/47f0036255982c7d5de69774faeda5fdd0022af5)
- dofm [`80081f3`](https://github.com/ruzickap/action-test/commit/80081f34b8fc7f44ea53026029c9453ad859b0ae)
- tv [`8ccc92c`](https://github.com/ruzickap/action-test/commit/8ccc92c6cf9892dcdf889bcf49e6528fffa68a93)
- aaa [`dc388c9`](https://github.com/ruzickap/action-test/commit/dc388c9db1c332ca4ce56bb49bafa9945952cf06)
- test [`89abf0c`](https://github.com/ruzickap/action-test/commit/89abf0c561f8a628a192e1f39c211b42ce0fb65e)
- gggg [`ce8e531`](https://github.com/ruzickap/action-test/commit/ce8e5317fa2a8cf4973e95b0bf6f705ce2edff99)
- ccc [`ea6938d`](https://github.com/ruzickap/action-test/commit/ea6938d24f2e7d508e1054eb1395ae75333c6af7)
- bvvv [`cbfd166`](https://github.com/ruzickap/action-test/commit/cbfd166998894c25a455460bfc954ff9b3ffa874)
- bbb [`16ddd50`](https://github.com/ruzickap/action-test/commit/16ddd503a6d9ed2879ac719a6955c4d2fcfe9877)
- iiaaa [`a26425d`](https://github.com/ruzickap/action-test/commit/a26425d3ae501a270f246a7b2ba1a3baa8dee131)

tests/run.sh
#!/bin/bash -eux

export INPUT_DEBUG="true"

# Test entrypoint.sh script

## Check differnet URLs types
export INPUT_CMD_PARAMS="--one-page-only --buffer-size=8192 --concurrency=10 --verbose --timeout=40"

export INPUT_URL="https://google.com"
../entrypoint.sh

export INPUT_URL="https://google.com:443"
../entrypoint.sh

export INPUT_URL="https://google.com:443/search"
../entrypoint.sh

## Test locally stored web pages (PAGES_PATH)
export INPUT_CMD_PARAMS="--skip-tls-verification --verbose"

export INPUT_PAGES_PATH="${PWD}"
export INPUT_URL="http://my-testing-domain.com/index2.html"
../entrypoint.sh

export INPUT_PAGES_PATH="${PWD}"
export INPUT_URL="https://my-testing-domain.com"
../entrypoint.sh

## Test docker image

docker build .. -t broken-link-checker-test-img

export INPUT_CMD_PARAMS="--one-page-only --buffer-size=8192 --concurrency=10 --verbose"
export INPUT_URL="https://google.com"
docker run --rm -t -e INPUT_DEBUG -e INPUT_URL -e INPUT_CMD_PARAMS broken-link-checker-test-img

export INPUT_CMD_PARAMS="--skip-tls-verification --verbose"
export INPUT_PAGES_PATH="${PWD}"
export INPUT_URL="https://my-testing-domain.com"
docker run --rm -t -e INPUT_DEBUG -e INPUT_URL -e INPUT_CMD_PARAMS -e INPUT_PAGES_PATH -v "$INPUT_PAGES_PATH:$INPUT_PAGES_PATH" broken-link-checker-test-img

export INPUT_CMD_PARAMS="--verbose"
export INPUT_PAGES_PATH="${PWD}"
export INPUT_URL="http://my-testing-domain.com/index2.html"
docker run --rm -t -e INPUT_DEBUG -e INPUT_URL -e INPUT_CMD_PARAMS -e INPUT_PAGES_PATH -v "$INPUT_PAGES_PATH:$INPUT_PAGES_PATH" broken-link-checker-test-img


tests/run_tests.sh
#!/bin/bash -eux

# Test entrypoint.sh script
export INPUT_DEBUG="true"


echo -e "\n\n*** Check differnet URLs types"
export INPUT_CMD_PARAMS="--one-page-only --buffer-size=8192 --concurrency=10 --verbose"

export INPUT_URL="https://google.com"
../entrypoint.sh

export INPUT_URL="https://google.com:443"
../entrypoint.sh

export INPUT_URL="https://google.com:443/search"
../entrypoint.sh


echo -e "\n\n*** Test locally stored web pages (PAGES_PATH)"
export INPUT_CMD_PARAMS="--skip-tls-verification --verbose"

export INPUT_PAGES_PATH="${PWD}"
export INPUT_URL="http://my-testing-domain.com/index2.html"
../entrypoint.sh

export INPUT_PAGES_PATH="${PWD}"
export INPUT_URL="https://my-testing-domain.com"
../entrypoint.sh


echo -e "\n\n*** Test docker image"
docker build .. -t broken-link-checker-test-img

export INPUT_CMD_PARAMS="--one-page-only --buffer-size=8192 --concurrency=10 --verbose"
export INPUT_URL="https://google.com"
docker run --rm -t -e INPUT_DEBUG -e INPUT_URL -e INPUT_CMD_PARAMS broken-link-checker-test-img

export INPUT_CMD_PARAMS="--skip-tls-verification --verbose"
export INPUT_PAGES_PATH="${PWD}"
export INPUT_URL="https://my-testing-domain.com"
docker run --rm -t -e INPUT_DEBUG -e INPUT_URL -e INPUT_CMD_PARAMS -e INPUT_PAGES_PATH -v "$INPUT_PAGES_PATH:$INPUT_PAGES_PATH" broken-link-checker-test-img

export INPUT_CMD_PARAMS="--verbose"
export INPUT_PAGES_PATH="${PWD}"
export INPUT_URL="http://my-testing-domain.com/index2.html"
docker run --rm -t -e INPUT_DEBUG -e INPUT_URL -e INPUT_CMD_PARAMS -e INPUT_PAGES_PATH -v "$INPUT_PAGES_PATH:$INPUT_PAGES_PATH" broken-link-checker-test-img


tests/failing_tests.sh
#!/bin/bash -ux

export INPUT_DEBUG="true"

echo -e "\n\n!!! Test when RUN_TIMEOUT is too tight"

export INPUT_CMD_PARAMS="--one-page-only --concurrency=1 --verbose"
export INPUT_RUN_TIMEOUT="1"
export INPUT_URL="https://google.com"
../entrypoint.sh
unset INPUT_RUN_TIMEOUT

echo -e "\n\n!!! Test nonexisting directory specified as PAGES_PATH"

export INPUT_PAGES_PATH="/non-existing-dir"
export INPUT_URL="https://google.com"
../entrypoint.sh

echo -e "\n\n!!! Test broken links by accessing wrong non existing domain"

export INPUT_PAGES_PATH="$PWD"
export INPUT_URL="https://non-existing-domain.com"
../entrypoint.sh

echo -e "\n\n!!! Test broken links by accessing non existing links"

export INPUT_PAGES_PATH="$PWD"
export INPUT_URL="https://my-testing-domain.com/index2.html"
../entrypoint.sh


entrypoint.sh
#!/usr/bin/env bash

export EXCLUDE="${INPUT_EXCLUDE}"

IFS=$'\n'

FIND_EXCLUDE=('-not' '-path' './node_modules/*')
for i in ${EXCLUDE[*]}; do
  echo "bbb ${i}"
  FIND_EXCLUDE+=('-not' '-path' "${i}")
done

FILE_EXTENSION="*.sh"

set -x

FIND_CALL=('find' '.' '-name' '*'"${FILE_EXTENSION}" "${FIND_EXCLUDE[@]}" '-exec' 'ls' '-la' '{}')

FIND_CALL+=(';')

find . -name "*.sh" -ls

"${FIND_CALL[@]}"

exit

# Command line parameters for muffet
export CMD_PARAMS="${INPUT_CMD_PARAMS:---buffer-size=8192 --concurrency=10}"
# Set path variable containing web pages
export PAGES_PATH=${INPUT_PAGES_PATH:-}
# URL to scan / check
export URL=${INPUT_URL:?}
# Domain where the web pages will be hosted (test.example.com), it will be stored in /etc/hosts
PAGES_DOMAIN=$( echo "${URL}" | awk -F[/:] '{print $4}' )
export PAGES_DOMAIN
# URI used by caddy to serve locally stored web pages (https://test.example.com)
PAGES_URI=$( echo "${URL}" | cut -d / -f 1,2,3 )
export PAGES_URI
# Maximum number of seconds that URL checker can be running
export RUN_TIMEOUT="${INPUT_RUN_TIMEOUT:-300}"
# Debug variable - enable by setting non-empty value
export DEBUG=${INPUT_DEBUG:-}

function print_error() {
  echo -e "\e[31m*** ERROR: ${1}\e[m"
}

function print_info() {
  echo -e "\e[36m*** INFO: ${1}\e[m"
}

# Remove all added files or changed /etc/hosts entry
function cleanup() {
  if [ -n "${PAGES_PATH}" ]; then
    # Manipulation with /etc/hosts using 'sed -i' doesn't work inside containers
    if ! grep -q docker /proc/1/cgroup ; then
      sudo sed -i "/127.0.0.1 ${PAGES_DOMAIN}  # Created by broken-link-checker/d" /etc/hosts
    fi
    [ -s "${CADDY_PIDFILE}" ] && sudo kill "$(cat "${CADDY_PIDFILE}")"
    [ -f "${CADDYFILE}" ] && rm "${CADDYFILE}"
  fi
}

function error_trap() {
  cleanup
  print_error "[$(date +'%F %T')] Something went wrong - see the errors above..."
}

################
# Main
################

trap error_trap ERR

[ -n "${DEBUG}" ] && set -x

# Install muffet if needed
if ! hash muffet &> /dev/null ; then
  MUFFET_URL=$(wget --quiet https://api.github.com/repos/raviqqe/muffet/releases/latest -O - | grep "browser_download_url.*muffet_.*_Linux_x86_64.tar.gz" | cut -d \" -f 4)
  wget --quiet "${MUFFET_URL}" -O - | sudo tar xzf - -C /usr/local/bin/ muffet
fi

# Install caddy if needed
if ! hash caddy &> /dev/null && [ -n "${PAGES_PATH}" ] ; then
  wget -qO- https://getcaddy.com | sudo bash -s personal > /dev/null
fi

# Use muffet in case of external URL check is required
if [ -z "${PAGES_PATH}" ] ; then
  # Run check
  print_info "[$(date +'%F %T')] Start checking: \"${URL}\""
  # shellcheck disable=SC2086
  timeout "${RUN_TIMEOUT}" muffet ${CMD_PARAMS} "${URL}"

else

  print_info "Using path \"${PAGES_PATH}\" as domain \"${PAGES_DOMAIN}\" with URI \"${PAGES_URI}\""

  # Die if the specified path which should contain local web pages doesn't exist
  if [ ! -d "${PAGES_PATH}" ]; then
    print_error "Path specified as 'INPUT_PAGES_PATH': '${PAGES_PATH}' doesn't exist!"
    exit 1
  fi

  # Add domain into /etc/hosts
  if ! grep -q "${PAGES_DOMAIN}" /etc/hosts ; then
    sudo bash -c "echo \"127.0.0.1 ${PAGES_DOMAIN}  # Created by broken-link-checker\" >> /etc/hosts"
  fi

  # Create caddy configuration to run web server using the domain set in PAGES_DOMAIN + /etc/hosts
  CADDYFILE=$( mktemp /tmp/Caddyfile.XXXXXX )
  CADDY_PIDFILE=$( mktemp -u /tmp/Caddy_pidfile.XXXXXX )
  echo -e "${PAGES_URI}\n root ${PAGES_PATH}\n tls self_signed" > "${CADDYFILE}"

  # Run caddy web server on the background
  sudo caddy -conf "${CADDYFILE}" -pidfile "${CADDY_PIDFILE}" -quiet &
  sleep 1

  # Run check
  print_info "[$(date +'%F %T')] Start checking: \"${URL}\""
  # shellcheck disable=SC2086
  timeout "${RUN_TIMEOUT}" muffet ${CMD_PARAMS} "${URL}"
  cleanup

fi

print_info "[$(date +'%F %T')] Checks completed..."
